## Data Properties

In [2]:
import pandas as pd
#Load the Races dataset
df_f1_races = pd.read_csv('races.csv')

df_f1_races.head()

raceId  year  round  circuitId                   name        date  \
0       1  2009      1          1  Australian Grand Prix  2009-03-29   
1       2  2009      2          2   Malaysian Grand Prix  2009-04-05   
2       3  2009      3         17     Chinese Grand Prix  2009-04-19   
3       4  2009      4          3     Bahrain Grand Prix  2009-04-26   
4       5  2009      5          4     Spanish Grand Prix  2009-05-10   

       time                                                url  
0  06:00:00  http://en.wikipedia.org/wiki/2009_Australian_G...  
1  09:00:00  http://en.wikipedia.org/wiki/2009_Malaysian_Gr...  
2  07:00:00  http://en.wikipedia.org/wiki/2009_Chinese_Gran...  
3  12:00:00  http://en.wikipedia.org/wiki/2009_Bahrain_Gran...  
4  12:00:00  http://en.wikipedia.org/wiki/2009_Spanish_Gran...

In [3]:
#Load the Races dataset
df_f1_results= pd.read_csv('f1_results.csv')

df_f1_results.head()

resultId  raceId  driverId  constructorId  number  grid  position  \
0         1      18         1              1    22.0     1       1.0   
1         2      18         2              2     3.0     5       2.0   
2         3      18         3              3     7.0     7       3.0   
3         4      18         4              4     5.0    11       4.0   
4         5      18         5              1    23.0     3       5.0   

  positionText  positionOrder  points  laps     time  milliseconds  \
0            1              1    10.0    58  34:50.6     5690616.0   
1            2              2     8.0    58    5.478     5696094.0   
2            3              3     6.0    58    8.163     5698779.0   
3            4              4     5.0    58   17.181     5707797.0   
4            5              5     4.0    58   18.014     5708630.0   

   fastestLap  rank fastestLapTime fastestLapSpeed  statusId  
0        39.0   2.0        01:27.5           218.3         1  
1        41.0   3.0        01:27.7         217.586         1  
2        41.0   5.0        01:28.1         216.719         1  
3        58.0   7.0        01:28.6         215.464         1  
4        43.0   1.0        01:27.4         218.385         1

In [4]:
print('the shape of the df_f1_races is', df_f1_races.shape)
print('the shape of the df_f1_results is', df_f1_results.shape)

the shape of the df_f1_races is (997, 8)
the shape of the df_f1_results is (23777, 18)


Now that we have a good understanding of our datasets, we will use the merge function to combine them both into one single dataset for better manipulation of both datasets. 

In [5]:
#Using the merge function to gather Results dataset and do a left join on the Races dataset
combineddata = df_f1_results.merge(df_f1_races, how = 'left', on='raceId')
combineddata.columns

Index(['resultId', 'raceId', 'driverId', 'constructorId', 'number', 'grid',
       'position', 'positionText', 'positionOrder', 'points', 'laps', 'time_x',
       'milliseconds', 'fastestLap', 'rank', 'fastestLapTime',
       'fastestLapSpeed', 'statusId', 'year', 'round', 'circuitId', 'name',
       'date', 'time_y', 'url'],
      dtype='object')

In [6]:
#Add section on renaming time_x and time_y 
combineddata.head()

resultId  raceId  driverId  constructorId  number  grid  position  \
0         1      18         1              1    22.0     1       1.0   
1         2      18         2              2     3.0     5       2.0   
2         3      18         3              3     7.0     7       3.0   
3         4      18         4              4     5.0    11       4.0   
4         5      18         5              1    23.0     3       5.0   

  positionText  positionOrder  points  ...  fastestLapTime fastestLapSpeed  \
0            1              1    10.0  ...         01:27.5           218.3   
1            2              2     8.0  ...         01:27.7         217.586   
2            3              3     6.0  ...         01:28.1         216.719   
3            4              4     5.0  ...         01:28.6         215.464   
4            5              5     4.0  ...         01:27.4         218.385   

   statusId  year  round circuitId                   name        date  \
0         1  2008      1         1  Australian Grand Prix  2008-03-16   
1         1  2008      1         1  Australian Grand Prix  2008-03-16   
2         1  2008      1         1  Australian Grand Prix  2008-03-16   
3         1  2008      1         1  Australian Grand Prix  2008-03-16   
4         1  2008      1         1  Australian Grand Prix  2008-03-16   

     time_y                                                url  
0  04:30:00  http://en.wikipedia.org/wiki/2008_Australian_G...  
1  04:30:00  http://en.wikipedia.org/wiki/2008_Australian_G...  
2  04:30:00  http://en.wikipedia.org/wiki/2008_Australian_G...  
3  04:30:00  http://en.wikipedia.org/wiki/2008_Australian_G...  
4  04:30:00  http://en.wikipedia.org/wiki/2008_Australian_G...  

[5 rows x 25 columns]

In [7]:
#Use .info() function to find characteristics of each column in the combined dataset
combineddata.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 23777 entries, 0 to 23776
Data columns (total 25 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   resultId         23777 non-null  int64  
 1   raceId           23777 non-null  int64  
 2   driverId         23777 non-null  int64  
 3   constructorId    23777 non-null  int64  
 4   number           23771 non-null  float64
 5   grid             23777 non-null  int64  
 6   position         13227 non-null  float64
 7   positionText     23777 non-null  object 
 8   positionOrder    23777 non-null  int64  
 9   points           23777 non-null  float64
 10  laps             23777 non-null  int64  
 11  time_x           6004 non-null   object 
 12  milliseconds     6003 non-null   float64
 13  fastestLap       5383 non-null   float64
 14  rank             5531 non-null   float64
 15  fastestLapTime   5383 non-null   object 
 16  fastestLapSpeed  5383 non-null   object 
 17  statusId    

### Statistical Properties: dataframe.describe()

Exploratory Data Analysis:

- EDA is an approach to analyzing data sets to summarize their main characteristics, often with visual methods

In [8]:
combineddata.describe()

resultId        raceId      driverId  constructorId        number  \
count  23777.000000  23777.000000  23777.000000   23777.000000  23771.000000   
mean   11889.481053    487.203937    226.515961      46.281785     16.965462   
std     6864.691322    269.904857    231.386102      56.174091     13.644798   
min        1.000000      1.000000      1.000000       1.000000      0.000000   
25%     5945.000000    273.000000     55.000000       6.000000      7.000000   
50%    11889.000000    478.000000    154.000000      25.000000     15.000000   
75%    17833.000000    718.000000    314.000000      57.000000     23.000000   
max    23781.000000    988.000000    843.000000     210.000000    208.000000   

               grid      position  positionOrder        points          laps  \
count  23777.000000  13227.000000   23777.000000  23777.000000  23777.000000   
mean      11.270303      7.782264      13.081591      1.601403     45.270598   
std        7.346436      4.745105       7.824711      3.665154     30.525404   
min        0.000000      1.000000       1.000000      0.000000      0.000000   
25%        5.000000      4.000000       7.000000      0.000000     20.000000   
50%       11.000000      7.000000      13.000000      0.000000     52.000000   
75%       17.000000     11.000000      19.000000      1.000000     66.000000   
max       34.000000     33.000000      39.000000     50.000000    200.000000   

       milliseconds   fastestLap         rank      statusId          year  \
count  6.003000e+03  5383.000000  5531.000000  23777.000000  23777.000000   
mean   6.303313e+06    41.061676    10.598807     18.242293   1987.675064   
std    1.721748e+06    17.156435     6.272457     26.380824     17.972231   
min    1.474899e+06     2.000000     0.000000      1.000000   1950.000000   
25%    5.442948e+06    29.000000     5.000000      1.000000   1975.000000   
50%    5.859428e+06    44.000000    11.000000     11.000000   1989.000000   
75%    6.495440e+06    53.000000    16.000000     16.000000   2003.000000   
max    1.509054e+07    78.000000    24.000000    136.000000   2017.000000   

             round     circuitId  
count  23777.00000  23777.000000  
mean       8.16987     21.735375  
std        4.78899     15.806595  
min        1.00000      1.000000  
25%        4.00000      9.000000  
50%        8.00000     18.000000  
75%       12.00000     32.000000  
max       21.00000     73.000000

***Exercise 1***
   - Import the titanic data set and find the datas shape, non-null values and data type for each column  and the overall statistical properties


In [20]:
titanic_data = pd.read_csv('titanic_data.csv')

In [21]:
print('the shape of the titanic dataset is', titanic_data.shape)

the shape of the titanic dataset is (1309, 14)


In [22]:
titanic_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1309 entries, 0 to 1308
Data columns (total 14 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   pclass     1309 non-null   int64 
 1   survived   1309 non-null   int64 
 2   name       1309 non-null   object
 3   sex        1309 non-null   object
 4   age        1309 non-null   object
 5   sibsp      1309 non-null   int64 
 6   parch      1309 non-null   int64 
 7   ticket     1309 non-null   object
 8   fare       1309 non-null   object
 9   cabin      1309 non-null   object
 10  embarked   1309 non-null   object
 11  boat       1309 non-null   object
 12  body       1309 non-null   object
 13  home.dest  1309 non-null   object
dtypes: int64(4), object(10)
memory usage: 143.3+ KB


In [23]:
titanic_data.describe()


pclass     survived        sibsp        parch
count  1309.000000  1309.000000  1309.000000  1309.000000
mean      2.294882     0.381971     0.498854     0.385027
std       0.837836     0.486055     1.041658     0.865560
min       1.000000     0.000000     0.000000     0.000000
25%       2.000000     0.000000     0.000000     0.000000
50%       3.000000     0.000000     0.000000     0.000000
75%       3.000000     1.000000     1.000000     0.000000
max       3.000000     1.000000     8.000000     9.000000

## Features

These are the various features present for each passenger on the ship:
- **Survived**: Outcome of survival (0 = No; 1 = Yes)...Categorical
- **Pclass**: Socio-economic class (1 = Upper class; 2 = Middle class; 3 = Lower class)...categorical
- **Name**: Name of passenger...Categorical
- **Sex**: Sex of the passenger....Categorical
- **Age**: Age of the passenger ....Numerical/Categorical
- **SibSp**: Number of siblings and spouses of the passenger aboard...Numerical

- **Parch**: Number of parents and children of the passenger aboard...?

- **Ticket**: Ticket number of the passenger....numerical
- **Fare**: Fare paid by the passenger...numerical
- **Cabin** Cabin number of the passenger ...categorical
- **Embarked**: Port of embarkation of the passenger (C = Cherbourg; Q = Queenstown; S = Southampton)....categorical

Lets explore the categorical features of the titanic data set
- We can use the value_count() method on each categorical feature to understand what categories the feature is made out of.

In [10]:
# import libraries
import pandas as pp

# Import titanic dataset
titanic_data = pd.read_csv("titanic_data.csv") 

In [11]:
# print the categories under the pclass feature and how many instances of each category are present in our data
print(titanic_data.pclass.value_counts())

3    709
1    323
2    277
Name: pclass, dtype: int64


- Note that there are 3 categories
- All categories are numbers, yet this is still a categorical feature

Conclusions:
- Most passangers in the titanic were lower class travelers.

***Exercise 2:***
   - Explore the other categorical features of our data set and draw conclusions about our data.


In [24]:
titanic_data.sex.value_counts()

male      843
female    466
Name: sex, dtype: int64

In [25]:
titanic_data.embarked.value_counts()

S    914
C    270
Q    123
?      2
Name: embarked, dtype: int64

## The Target Variable

Which column  should you use as your target variable from the given F1 data?
- You might say position, but after further evaluation you will realize that the 'position' column has missing values

In [13]:
# Extract the number of empty values in your selected target value
combineddata['position'].isna().sum()

10550

- There are 10550 empty records in this column, the data is incomplete and will not allow us to make good predictions of the final position of each racer.

Furtunately, we can try another column or gain more data:

- Note that there are no empty values in the positionOrder column, this would then be a better target value to make predictions

In [14]:
# Extract the number of empty values from the 'positionOrder' column
combineddata['positionOrder'].isna().sum()

0

***What type of target will we be predicting on?***

- If you decide to predict weather a driver will win the race or not, the correct model to use would be a classification model 
    - The classes for which you will predict are 1 and 0
    - 1 = driver wins the race
    - 0 = driver does not win the race

- Note that the prediction in a classification model can also be called a label.
    - In this case we have two labels 1 and 0

***In the case of the F1 dataset...***
- We must now create a target variable based on the position order column
- If the value of 'positionOrder' = 1, the target column will equal 1
- If the value of 'positionOrder' != 1, then the target column will equal 0

In [15]:
# Create a 'label' column based on the position order
combineddata['label'] = combineddata['positionOrder']==1

In [16]:
combineddata

resultId  raceId  driverId  constructorId  number  grid  position  \
0             1      18         1              1    22.0     1       1.0   
1             2      18         2              2     3.0     5       2.0   
2             3      18         3              3     7.0     7       3.0   
3             4      18         4              4     5.0    11       4.0   
4             5      18         5              1    23.0     3       5.0   
...         ...     ...       ...            ...     ...   ...       ...   
23772     23777     988       842              5    10.0    17      16.0   
23773     23778     988       828             15     9.0    19      17.0   
23774     23779     988       840              3    18.0    15      18.0   
23775     23780     988       832              4    55.0    12       NaN   
23776     23781     988       817              9     3.0     4       NaN   

      positionText  positionOrder  points  ...  fastestLapSpeed statusId  \
0                1              1    10.0  ...            218.3        1   
1                2              2     8.0  ...          217.586        1   
2                3              3     6.0  ...          216.719        1   
3                4              4     5.0  ...          215.464        1   
4                5              5     4.0  ...          218.385        1   
...            ...            ...     ...  ...              ...      ...   
23772           16             16     0.0  ...          192.542       11   
23773           17             17     0.0  ...          193.057       11   
23774           18             18     0.0  ...          195.402       11   
23775            R             19     0.0  ...           193.41       36   
23776            R             20     0.0  ...          194.579        9   

       year  round  circuitId                   name        date    time_y  \
0      2008      1          1  Australian Grand Prix  2008-03-16  04:30:00   
1      2008      1          1  Australian Grand Prix  2008-03-16  04:30:00   
2      2008      1          1  Australian Grand Prix  2008-03-16  04:30:00   
3      2008      1          1  Australian Grand Prix  2008-03-16  04:30:00   
4      2008      1          1  Australian Grand Prix  2008-03-16  04:30:00   
...     ...    ...        ...                    ...         ...       ...   
23772  2017     20         24   Abu Dhabi Grand Prix  2017-11-26  13:00:00   
23773  2017     20         24   Abu Dhabi Grand Prix  2017-11-26  13:00:00   
23774  2017     20         24   Abu Dhabi Grand Prix  2017-11-26  13:00:00   
23775  2017     20         24   Abu Dhabi Grand Prix  2017-11-26  13:00:00   
23776  2017     20         24   Abu Dhabi Grand Prix  2017-11-26  13:00:00   

                                                     url  label  
0      http://en.wikipedia.org/wiki/2008_Australian_G...   True  
1      http://en.wikipedia.org/wiki/2008_Australian_G...  False  
2      http://en.wikipedia.org/wiki/2008_Australian_G...  False  
3      http://en.wikipedia.org/wiki/2008_Australian_G...  False  
4      http://en.wikipedia.org/wiki/2008_Australian_G...  False  
...                                                  ...    ...  
23772  https://en.wikipedia.org/wiki/2017_Abu_Dhabi_G...  False  
23773  https://en.wikipedia.org/wiki/2017_Abu_Dhabi_G...  False  
23774  https://en.wikipedia.org/wiki/2017_Abu_Dhabi_G...  False  
23775  https://en.wikipedia.org/wiki/2017_Abu_Dhabi_G...  False  
23776  https://en.wikipedia.org/wiki/2017_Abu_Dhabi_G...  False  

[23777 rows x 26 columns]

In [17]:
# Find the class distribution
combineddata.groupby('label')['label'].count()

label
False    22798
True       979
Name: label, dtype: int64

 - Note that the negative class has more instances that the positive class

***Exercise 3***

     1) What type of model do we need to use in order to predict whether or not a passanger survived the titanic based on the given data? (regression/classification) 
     2) Find the number of positive and negative labels in the titanic data

In [18]:
 # A1: Classification

In [26]:
titanic_data.survived.value_counts()

0    809
1    500
Name: survived, dtype: int64

    3) Based on the distribution of negative and positive samples, what can you say about our future predictions

In [19]:
#

***Exercise 4***
Explore the number of the 'age' feature from the titanic data set.

    1) How many different ages exist in our data set
    2) How many people are 21 years old?
    3) Are there any unexpected values in our dataset?
    4) Remove all samples where we have an unknown age
    

In [27]:
titanic_data.age.value_counts()

?         263
24         47
22         43
21         41
30         40
         ... 
0.3333      1
23.5        1
67          1
0.1667      1
38.5        1
Name: age, Length: 99, dtype: int64

In [28]:
titanic_data.loc[titanic_data.age == '21', 'age'].count()


41

In [29]:
titanic_data = titanic_data.replace({'?': None})

In [30]:
titanic_data

pclass  survived                                             name  \
0          1         1                    Allen, Miss. Elisabeth Walton   
1          1         1                   Allison, Master. Hudson Trevor   
2          1         0                     Allison, Miss. Helen Loraine   
3          1         0             Allison, Mr. Hudson Joshua Creighton   
4          1         0  Allison, Mrs. Hudson J C (Bessie Waldo Daniels)   
...      ...       ...                                              ...   
1304       3         0                             Zabour, Miss. Hileni   
1305       3         0                            Zabour, Miss. Thamine   
1306       3         0                        Zakarian, Mr. Mapriededer   
1307       3         0                              Zakarian, Mr. Ortin   
1308       3         0                               Zimmerman, Mr. Leo   

         sex     age  sibsp  parch  ticket      fare    cabin embarked  boat  \
0     female      29      0      0   24160  211.3375       B5        S     2   
1       male  0.9167      1      2  113781    151.55  C22 C26        S    11   
2     female       2      1      2  113781    151.55  C22 C26        S  None   
3       male      30      1      2  113781    151.55  C22 C26        S  None   
4     female      25      1      2  113781    151.55  C22 C26        S  None   
...      ...     ...    ...    ...     ...       ...      ...      ...   ...   
1304  female    14.5      1      0    2665   14.4542     None        C  None   
1305  female    None      1      0    2665   14.4542     None        C  None   
1306    male    26.5      0      0    2656     7.225     None        C  None   
1307    male      27      0      0    2670     7.225     None        C  None   
1308    male      29      0      0  315082     7.875     None        S  None   

      body                        home.dest  
0     None                     St Louis, MO  
1     None  Montreal, PQ / Chesterville, ON  
2     None  Montreal, PQ / Chesterville, ON  
3      135  Montreal, PQ / Chesterville, ON  
4     None  Montreal, PQ / Chesterville, ON  
...    ...                              ...  
1304   328                             None  
1305  None                             None  
1306   304                             None  
1307  None                             None  
1308  None                             None  

[1309 rows x 14 columns]